include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h

```diff
@@ -154,17 +274,201 @@ class PreconditionerML : public PreconditionerBase<double>
+      if(ml_data.higher_order_elements)
+      {
+        parameter_list.set("aggregation: type", "Uncoupled");
+      }
+    }
+    else
+    {
+      ML_Epetra::SetDefaults("NSSA", parameter_list);
+      parameter_list.set("aggregation: type", "Uncoupled");
+      parameter_list.set("aggregation: block scaling", true);
+    }
+
+    parameter_list.set("smoother: type", ml_data.smoother_type);
+    parameter_list.set("coarse: type", ml_data.coarse_type);
+
+// Force re-initialization of the random seed to make ML deterministic
+// (only supported in trilinos >12.2):
+#  if DEAL_II_TRILINOS_VERSION_GTE(12, 4, 0)
+    parameter_list.set("initialize random seed", true);
+#  endif
+
+    parameter_list.set("smoother: sweeps", static_cast<int>(ml_data.smoother_sweeps));
+    parameter_list.set("cycle applications", static_cast<int>(ml_data.n_cycles));
+    if(ml_data.w_cycle == true)
+    {
+      parameter_list.set("prec type", "MGW");
+    }
+    else
+    {
+      parameter_list.set("prec type", "MGV");
+    }
+
+    parameter_list.set("smoother: Chebyshev alpha", 10.);
+    parameter_list.set("smoother: ifpack overlap", static_cast<int>(ml_data.smoother_overlap));
+    parameter_list.set("aggregation: threshold", ml_data.aggregation_threshold);
+
+    // Minimum size of the coarse problem, i.e. no coarser problems
+    // smaller than `coarse: max size` are constructed.
+    parameter_list.set("coarse: max size", 2000);
+
+    // This extends the settings in deal::PreconditionAMG::AdditionalData::set_parameters().
+    parameter_list.set("repartition: enable", 1);
+    parameter_list.set("repartition: max min ratio", 1.3);
+    parameter_list.set("repartition: min per proc", 300);
+    parameter_list.set("repartition: partitioner", "Zoltan");
+    parameter_list.set("repartition: Zoltan dimensions", static_cast<int>(dim));
+
+    if(ml_data.output_details)
+    {
+      parameter_list.set("ML output", 10);
+    }
+    else
+    {
+      parameter_list.set("ML output", 0);
+    }
+
+    return parameter_list;
+  }
+
+  std::vector<VectorType>
+  get_constant_modes()
+  {
+    unsigned int n_constant_modes = dealii::numbers::invalid_unsigned_int;
+    if(operator_type == AMGOperatorType::ScalarLaplace)
+    {
+      n_constant_modes = 1;
+    }
+    else if(operator_type == AMGOperatorType::VectorLaplace)
+    {
+      n_constant_modes = dim;
+    }
+    else if(operator_type == AMGOperatorType::Elasticity)
+    {
+      n_constant_modes = dim == 1 ? 1 : (dim == 2 ? 3 : 6);
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("AMGOperatorType not defined"));
+    }
+
+    std::vector<VectorType> constant_modes(n_constant_modes);
+
+    // The AMG Preconditioner might be used as a coarse-level solver/preconditioner within
+    // multigrid.
+    bool const on_coarse_grid = pde_operator.get_matrix_free().get_dof_handler().has_level_dofs();
+
+    for(unsigned int i = 0; i < n_constant_modes; ++i)
```

Comment:
NOTE TO @richardschu :
this cannot work if the `dof_index` is not 0 (e.g. pressure in Navier-Stokes).
You need something along the lines of   
`

      // Fill vector with rigid body modes.
      unsigned int const dof_index = pde_operator.get_dof_index();
	  unsigned int const n_components = pde_operator.get_matrix_free().get_dof_handler(dof_index).get_fe().n_components();
	  dealii::ComponentMask component_mask(n_components, true);
	  if(n_components == 1)
	  {
		  ConstantModes<dim, 1> const mode_generator(operator_type, i);
		  dealii::VectorTools::interpolate(mapping,
										   pde_operator.get_matrix_free().get_dof_handler(dof_index),
										   mode_generator,
										   constant_modes[i],
										   component_mask,
										   operating_on_fine_else_coarsest_level ?
											 dealii::numbers::invalid_unsigned_int :
											 0);
	  }
	  else if(n_components == dim)
	  {
		  ConstantModes<dim, dim> const mode_generator(operator_type, i);
		  dealii::VectorTools::interpolate(mapping,
										   pde_operator.get_matrix_free().get_dof_handler(dof_index),
										   mode_generator,
										   constant_modes[i],
										   component_mask,
										   operating_on_fine_else_coarsest_level ?
											 dealii::numbers::invalid_unsigned_int :
											 0);
	  }
`